# n차원 배열에서 최댓값 index 뽑아내기
> numpy array 배열에서 최댓값을 찾고, 그 index를 뽑아내는 함수

In [1]:
# 라이브러리 불러오기
import numpy as np 
import random

sample_li2 = list(range(1000)) # 리스트 생성
sample_li2 = random.sample(sample_li2, 500) # 1000개의 숫자 중 무작위로 500개 추출해 리스트로 만듦
sample_li2 = np.array(sample_li2) # numpy array로 변환
sample_li2

array([932, 377, 695, 365, 573, 403, 859, 949, 486, 307, 939, 676, 852,
       409,  67, 326, 605,   9, 125, 344, 810, 463, 839, 120, 539, 624,
       612, 587, 671,  33, 865,  11, 783, 890, 107, 541, 937, 774, 515,
       760, 640, 778, 466, 849, 480, 370, 793, 667, 203, 329, 770, 265,
       781, 775, 351, 809, 836, 324, 522, 928, 464, 535, 220,  31, 534,
       134, 824, 982, 339, 841, 888, 964, 489,  32,  12, 876, 921, 806,
       677, 899, 787, 335, 399, 772, 967, 250, 462, 330, 965, 794, 649,
       131,  72, 137, 215, 258, 448, 231, 180, 808, 328, 566,  47, 311,
       990, 594, 609, 882, 940, 776, 236,  44, 304,  37, 705, 298,  19,
       596, 272, 346, 626, 762,  46, 431, 288, 146, 270, 504, 740, 564,
       581, 975, 183, 286, 700, 293,   6, 988,  64,  88, 678, 771, 968,
        56, 546, 126, 398, 537, 754, 327, 561, 563, 523, 424, 156, 862,
       273, 920, 684, 763, 108, 697, 866, 274, 555, 593,  97, 383, 840,
       488, 277,  16, 884, 202, 983, 823, 269, 744, 367, 133, 64

In [2]:
# 최댓값 확인하기
max_li = np.max(sample_li2)
max_li

998

최댓값은 998입니다. 이러한 최댓값의 index는 np.argmax를 통해 뽑아낼 수 있습니다.

In [4]:
# 최댓값의 index 뽑아내기 
max_idx = np.argmax(sample_li2)
max_idx

208

index가 208이라 나왔습니다. np.where를 통해 해당 index 값이 맞는지 확인해보겠습니다.

In [5]:
# index 값 확인하기
max_idx = np.where(sample_li2==998) # 값이 998인 item의 index를 반환
max_idx

(array([208], dtype=int64),)

index 값이 일치하는 것을 확인할 수 있습니다. 이번에는 2차원 배열에서도 적용할 수 있는지 보겠습니다. 무작위 배열을 생성하기 위해 이번에는 먼저 리스트를 작성하고 변환하는 것이 아닌, random.random 함수를 사용하겠습니다.

In [6]:
sample_li = np.random.random(size=(2,5))
sample_li

array([[0.59543381, 0.22452867, 0.98197496, 0.2707333 , 0.57917919],
       [0.01624734, 0.61057729, 0.86482696, 0.50893321, 0.96918831]])

In [7]:
max_idx = np.argmax(sample_li)
max_idx

2

뭔가 이상합니다. 3차원 배열의 예로도 실험해보겠습니다. 이번엔 최댓값을 미리 구해보죠.

In [8]:
sample_li = np.random.random(size=(3,5))
sample_li

array([[0.29675798, 0.59401098, 0.53726415, 0.02265291, 0.60105062],
       [0.49993979, 0.05902082, 0.10896415, 0.20982479, 0.86586717],
       [0.56235554, 0.97958424, 0.89860874, 0.15069784, 0.01653617]])

In [9]:
max_li = np.max(sample_li)
max_li

0.9795842390361786

In [10]:
max_idx = np.argmax(sample_li)
max_idx

11

다차원 배열에서 index가 하나의 값으로 나오는 건 이상한 일입니다. np.where를 써서 확인해보겠습니다. 

In [12]:
max_idx = np.where(sample_li==max_li)
max_idx

(array([2], dtype=int64), array([1], dtype=int64))

이건 무슨 뜻일까요? index번호가 2인 행의 두 번째 item이라는 것일까요? 확인해보겠습니다.

In [13]:
sample_li[2][1]

0.9795842390361786

추측이 맞음을 알 수 있습니다. 그러나 왜 np.argmax의 값이 11로 나왔는지는 아직 모르겠습니다. 이를 알아내기 위해 sample_li를 list로 바꾸어 보겠습니다. 해당 작업에는 np.ravel 함수와 tolist 함수가 함수가 필요합니다.

In [15]:
arrtoli = np.ravel(sample_li) # 3차원 배열을 1차원으로 변환
arrtoli

array([0.29675798, 0.59401098, 0.53726415, 0.02265291, 0.60105062,
       0.49993979, 0.05902082, 0.10896415, 0.20982479, 0.86586717,
       0.56235554, 0.97958424, 0.89860874, 0.15069784, 0.01653617])

In [16]:
arrtoli = arrtoli.tolist() # 1차원 배열을 list로 변환
arrtoli

[0.2967579756845099,
 0.5940109770320633,
 0.5372641513440417,
 0.022652906746921664,
 0.6010506225010077,
 0.49993978644722037,
 0.059020823278518986,
 0.10896414832087686,
 0.2098247895511336,
 0.8658671701483164,
 0.562355543247731,
 0.9795842390361786,
 0.8986087385025636,
 0.15069783956399085,
 0.01653617130651963]

In [18]:
max_li = max(arrtoli) # list에서 최댓값 뽑아내기
max_li = arrtoli.index(max_li) # 최댓값의 index 찾기
max_li

11

결과에서 알 수 있듯이, np.argmax는 배열 내 최댓값의 index를 반환하지만 n차원 배열도 1차원 배열처럼 간주하여 index 값을 돌려줍니다. 정확한 index를 알고 싶다면 np.where를 쓰는 것이 더 도움이 될 것입니다.